### ISSUE TITLE
Creator payout data format - extended

### ISSUE THREAD

**Issue by zeeshanakram3** [Joystream/joystream#3163]
This issue is an extension of #2943 and provides a rough solution to problems addressed in the said issue.

The content here also assemble some ideas from the following:

* #2365 
* #2420
* https://github.com/Joystream/joystream/issues/2943#issuecomment-1026701296
* [Claiming Code](https://github.com/Joystream/joystream/blob/olympia/runtime-modules/content/src/lib.rs#L1412) and relevant test [functions](https://github.com/Joystream/joystream/blob/6b95450f750d6adc233933b2578951e1eff1bb85/runtime-modules/content/src/tests/fixtures.rs#L1544)

Let
`c_i`  = channel id that is owed some reward
`p_i`  = cumulative reward owed to channel with id `c_i`
`m_i` = some commitment that justifies the amount `p_i` for `c_i`
`d_i`  = description explaining the rationale for for reward or deduction for `c_i`

The set `S` of all payout claims where each claim consists of `(c_i, p_i, m_i, d_i)` is
:

1. Orderable on channel Id.
1. Unique - only one payout entry `{c_i, p_i, m_i, d_i}` exists for each channel `c_i` (emphasizing on this because there are multiple payout entries in the Set for same channel Id in implemented [test cases](https://github.com/Joystream/joystream/blob/6b95450f750d6adc233933b2578951e1eff1bb85/runtime-modules/content/src/tests/fixtures.rs#L1634)

If the above two properties are satisfied then `S` can be represented in the form of the following data structures:

## Ordered Set
The time complexity for:

* Searching the existence of payout for `c_i` in S is _`O(log n)`_
* Updating payout for some already existing channel Id `c_i` to new reward `p_new` is _`O(log n)`_
* Inserting payout for some new channel Id `c_new` is _`O(n)`_

## Binary Search Tree
The time complexity for:

* Searching the existence of payout for `c_i` in S is _`O(log n)`_ (average case)
* Updating payout for some already existing channel Id `c_i` to new reward `p_new` is _`O(log n)`_ (average case)
* Inserting payout for some new channel Id `c_new` is _`O(n)`_ (average case)

However, for unbalanced tree complexity can in order of `n` so AVL tree seems better alternative.

## AVL Tree
The time complexity for:

* Searching the existence of payout for `c_i` in S is _`O(log n)`_
* Updating payout for some already existing channel Id `c_i` to new reward `p_new` is _`O(log n)`_
* Inserting payout for some new channel Id `c_new` is _`O(log n)`_

## Code
The following functions would be implemented on TS side:

`function insertChannelsPayoutSet(payout: (c, p, m, d)): { ... }` 
`function retrieveChannelPayout(channelId: number): { ... }`

If the cardinality of a payout set `S` is `n` then space complexity of set `H` containing all Merkle branches of each claim in `S` would be _`O(log n)`_; more precisely ~2n 

So, in short, there are two sets - one containing payouts and the other containing Merkle branches. Also, actual proofs from the main raw payload(`S`) can be calculated and then compared against their resected values in `H`

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201958687417145/1201958614651845) by [Unito](https://www.unito.io)

**Comment by bedeho** [comment]
# Feedback 
- The structure of this writeup was actually a bit confusing, not sure how all the different trees are relating to each other or what the actual proposal is, and this does not appear to actually address the core question of proposing an actual data _format_.
- There is no need for doing computation on the data format, as you are doing with updating and inserting. The raw data will live in some entirely different representation, which will be where any updating will occur, and then only the final output `F` to be used for the on-chain commitment is relevant to this data format.
- Any solution involving searching `O(log N)` read operations will be too slow because of latency of reading from remote data source - as clients cannot download all of `F` locally for these reads. Say if we have 10M channels, then there has to be 23 request-response round trips, if each takes 1200ms e2e (not unlikely with all the random offset reads required), this is almost 30s, which is quite rough. It also will involve. 

# Proposal 

What about something simple like the following below. If it make sense, try to propose an exact schema using protocol buffers, and make sure the data you pick will work well with the native on-chain data types.

## Payload Format

Some sort of versioning information may make sense here, but protocol buffers may already natively support this in some way, I am not sure.

### Header

This can be downloaded fully by all clients that wish to inspect their current status and possibly cash out, as it likely will be quite small.

- P: byte length of entire payload
- H: byte length of header
- N: number of channels
- N times: serialized `c_i||offset_i`, where `offset_i` is byte offset w.r.t. start of payload where record for channel `c_i` exists. May as well be ordered in terms of `i`, but is not important if entire header is usually downloaded.
  
###  Body

This is not downloaded by clients in full, instead a client will just download the relevant record directly using he header lookup table.

- N times: serialized `c_i||p_i||proof_i||d_i` where `proof_i` is the serialized merkle branch described in section below.

## Cashout tree and proof generation

Let `CASHOUT_VECTOR = [c_1||p_1, ... , c_N||p_N]`, and `R` be the merkle root generated for cashouts over `CASHOUT_VECTOR`, which is what is going to be comitted to on-chain, and let `proof_i` be the merkle branch for leaf `c_i||p_i`. 

## Cashout council proposal

> This is not part of scope for proposal, just adding for clarity.

The council commits to the following in their proposal
- Some rationale for payout
- `R`
- Payload hash + size

The side-effects are
- `R` is set as new payout state root in content directory
- A new data object upload is facilitated in the data directory, based on payloadh hash + size, which allows someone to actually upload the raw payload to the storage system council bag.

**Comment by zeeshanakram3** [comment]
> ### Cashout council proposal
> ...
> The council commits to the following in their proposal

Form the three things mentioned only one(`R`) is being committed on-chain as visible in the signature of `update_commitment` extrinsic.

❌   Some rationale for payout

✅ `R`
   
❌  Payload hash + size

```rs
pub fn update_commitment(
    origin,
    new_commitment: <T as frame_system::Trait>::Hash,
) {
```

Will other two be updated by some different extrinsic or `update_commitment` function will include them?

**Comment by bedeho** [comment]
`update_commitment` is not the full store, this is a Root origin dispatch which will be called _from a proposal_, this proposal will have multiple parameters, those being hash, size and rationale.
